File: SatinalmaPro.Shared/Procurement/FirestoreQueryBuilder.cs
namespace SatinalmaPro.Shared.Procurement;

/// <summary>
/// Firestore REST structured query veya log için insan okunur sorgu özeti.
/// Masaüstü HTTP istemcisi bu özeti referans alarak sorgu oluşturur.
/// </summary>
public static class FirestoreQueryBuilder
{
    public static string Describe(FirestoreFilterSpec spec)
    {
        var parts = new List<string> { $"collection={spec.Collection}" };

        if (spec.StatusIn.Count > 0)
            parts.Add($"status in [{string.Join(", ", spec.StatusIn)}]");

        if (!string.IsNullOrWhiteSpace(spec.RequesterUidEquals))
            parts.Add($"requesterUid == {spec.RequesterUidEquals}");

        if (spec.UrgentFirst)
            parts.Add("order: priority DESC (urgent first)");

        if (spec.OrderBy.Count > 0)
        {
            var order = string.Join(", ", spec.OrderBy.Select(o =>
                $"{o.Field} {(o.Descending ? "DESC" : "ASC")}"));
            parts.Add($"orderBy: {order}");
        }

        if (spec.ReadOnly)
            parts.Add("readOnly=true");

        if (spec.RequiresReturnFlag)
            parts.Add("hasReturnFlag == true");

        return string.Join(" · ", parts);
    }

    public static FirestoreFilterSpec? ForTab(ProcurementTab tab, string? role, string? currentUid) =>
        TabFilterManager.GetQuerySpec(tab, role, currentUid);
}

File: SatinalmaPro.Android/app/src/main/java/com/satinalmapro/shared/filter/FirestoreQueryBuilder.kt
package com.satinalmapro.shared.filter

import com.google.firebase.firestore.FirebaseFirestore
import com.google.firebase.firestore.Query

/**
 * [FirestoreFilterSpec] → Firebase Firestore [Query] dönüştürücü.
 */
object FirestoreQueryBuilder {

    fun build(db: FirebaseFirestore, spec: FirestoreFilterSpec): Query {
        var query: Query = db.collection(spec.collection)

        when (spec.statusIn.size) {
            0 -> Unit
            1 -> query = query.whereEqualTo("status", spec.statusIn.first())
            else -> query = query.whereIn("status", spec.statusIn)
        }

        if (!spec.requesterUidEquals.isNullOrBlank()) {
            query = query.whereEqualTo("requesterUid", spec.requesterUidEquals)
        }

        spec.orderBy.forEach { order ->
            query = if (order.descending) {
                query.orderBy(order.field, Query.Direction.DESCENDING)
            } else {
                query.orderBy(order.field, Query.Direction.ASCENDING)
            }
        }

        return query
    }

    fun buildForTab(
        db: FirebaseFirestore,
        tab: ProcurementTab,
        role: String?,
        currentUid: String?
    ): Query? {
        val spec = TabFilterManager.getQuerySpec(tab, role, currentUid) ?: return null
        return build(db, spec)
    }
}
